keps/sig-cloud-provider/20200928-extend-konnectivity-for-both-directions.md

```diff
@@ -0,0 +1,194 @@
+* Define a solution involving less than one agent per node. 
+
+## Definitions
+
+* **Master Network** An IP reachable network space which contains the master components, such as Kubernetes API Server, Connectivity Proxy and ETCD server.
+* **Node Network** An IP reachable network space which contains all the clusters Nodes, for alpha. Worth noting that the Node Network may be completely disjoint from the Master network. It may have overlapping IP addresses to the Master Network or other means of network isolation. Direct IP routability between cluster and master networks should not be assumed. Later versions may relax the all node requirement to some.
+* **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+* **Konnectivity Server** The proxy server which runs in the master network. It has a secure channel established to the cluster network. It could work on either a HTTP Connect mechanism or gRPC. If the former it would expose a gRPC interface to KAS to provide connectivity service. If the latter it would use standard HTTP Connect. Formerly known as the Network Proxy Server.
+* **Konnectivity Agent** A proxy agent which runs in the node network for establishing the tunnel. Formerly known as the Network Proxy Agent.
+* **Flat Network** A network where traffic can be successfully routed using IP. Implies no overlapping (i.e. shared) IPs on the network.
+
+## Proposal
+
+Currently the Konnectivity Server is accepting requests from the KAS either with the gRPC or the HTTP Connect interfaces and is taking care of forwarding the traffic to the Konnectivity Agent using the previously established connections (initiated by the Agents). 
+
+In order to enable traffic from Kubelets and Pods running on Master Network, the Konnectivity Agents have to expose an endpoint that will be listening on a specific port for each of the destinations on the Master Network. As opposed to the traffic flowing from the Master Network to the Node Network, the Konnectivity Agent should act transparently: From a Kubelets or Pods standpoint, the Konnectivity Agent should be the final destination instead of acting as a proxy. 
+
+The reason why we do not intend to use the same strategy used in the other direction is that we do not have control over the clients using the Kubernetes default service to communicate with the KAS.
+
+### Traffic Flow
+
+```
+client =TCP=> (:6443) agent GRPC=> server =TCP=> KAS(:6443)
+                         |            ^
+                         |   Tunnel   |
+                         +------------+
+```
+
+The agent listens for TCP connections at a specific port for each configured destination. When a connection request is received by the Konnectivity Agent the following happens:
+1. A GRPC DIAL_REQ message is sent to the Konnectivity server containing the destination address associated with the current port.
```

Comment:
Yes, this is what we had in mind. I can make this explicit if you think it's necessary.

I'm not sure it would be easy to send the DIAL_REQ at SYN or SYN/ACK phase. How could we do it? 
As far as I know [TCPListener.Accept](https://golang.org/pkg/net/#TCPListener.Accept) method returns the connection once the TCP handshake is over.

BTW do you foresee any advantage in sending the DIAL_REQ before the TCP handshake is over?